### PR TITLE
Timeout of diff algrorithm

### PR DESCRIPTION
**Pull request by guibou**
But this partially fix https://github.com/NorfairKing/sydtest/issues/92 by introducing a timeout of 2s on the diff computatation.

The 2s is hardcoded yada yada

Don't merge, I will detail the commit message later and discuss the `unsafePerformIO` trick (if you accept it ;)

### CODE REVIEW

File: sydtest/src/Test/Syd/Output.hs [hunk 1]
```diff
@@ -463,8 +469,36 @@ splitChunksIntoLines =
...
-  let diff = V.toList $ getTextDiff (T.pack actual) (T.pack expected)
-      -- Add a header to a list of lines of chunks
+  let diffM = unsafePerformIO $ do
+        let diff = getTextDiff (T.pack actual) (T.pack expected)
+        -- we want to evaluate the diff in order to ensure that its
+        -- computation happen in the timeout block
+        -- and is not instead lazyly passed to the caller.
+        --
+        -- The safe option would be to @force diff@ to normal form, however
+        -- there are no 'NFData' instances on 'Diff Text'.
```
Comment (NorfairKing) [inline sydtest/src/Test/Syd/Output.hs:479]: I own that code so we can go put them there.

Comment (NorfairKing) [reply sydtest/src/Test/Syd/Output.hs:479]: https://github.com/NorfairKing/fast-myers-diff/blob/6cee293f3f32c443295e589330cf30a46ada0045/fast-myers-diff/src/Myers/Diff.hs#L52

Comment (guibou) [reply sydtest/src/Test/Syd/Output.hs:479]: https://github.com/NorfairKing/fast-myers-diff/pull/1 MR is here, I let you merge (if you accept) and I'll rework the current MR here.

(I'm really not a friend of hackage bounds, but looks like a release of `fasty-myers-diff` will be necessary in order for `sydtest` to be released with new bounds)

Comment (guibou) [reply sydtest/src/Test/Syd/Output.hs]: MR upgraded, it now uses `force` and `fast-myres-diff > 0.0.1`.

**Comment by guibou** [comment]
@NorfairKing I'm unsure about the `unsafePerformIO` there. It it ugly, but changing that to a true `IO` would force us to bubble up the timeout up until `printOutputSpecForest settings results` which will be a lot of `IO`.

Or the logic can be changed so the timeout is only done inside `printOutputSpecForest` and once for the complete test result, and if the initial `renderResultReport settings tc results` does not returns in less than `x` seconds, we cancel and rerun it again with a new setting which disable diff computation.

It will be simpler in term of code, with less `IO` (or `unsafePerformIO`) involved, and maybe having a global timeout could be a bit better because on a large test suite, having `2s` of diff per failure could quickly accumulate to a lot of time.

However, having a global timeout forces to reduce the complet output in memory (e.g. no more lazy text).

That's why I think the `unsafePerformIO` here is simpler. What's your point of view?

**Comment by NorfairKing** [comment]
@guibou I proposed an alternative approach for avoiding `unsafePerformIO`.
This will be a breaking change for some helper functions in `Test.Syd.Expectation` (stringsNotEqualButShouldHaveBeenEqual, for example) but I think I'd prefer that over `unsafePerformIO`. What do you think?

**Comment by guibou** [comment]
@NorfairKing I'm unsure were I should look to see your alternate approach ?

**Comment by NorfairKing** [comment]
> @NorfairKing I'm unsure were I should look to see your alternate approach?

I didn't press "complete review", it should be visible now.

**Comment by NorfairKing** [comment]
@guibou The `fast-myers-diff` change is in and released so we can continue here.

**Comment by guibou** [comment]
Hi @NorfairKing,

sorry, got occupied by life.

I refactored the code (still need some comment I guess), with the following properties:

- No more `unsafePerformIO` as you suggested, the diff is computed in `shouldBe`
- Because the test is computed in `shouldBe`, the timing is part of the test timing, not hidden in the global test suite. Note that however, the test diff is produced at that time and hence stays in memory, and the diff is also recomputed as many time as we retry, so it can take up to 6s by default. Maybe we could reduce the default from 2s to something else. (To be honest, when the test is taking more than a few ms, the diff is most of the time completly useless).
- Because the input string are forced in the `shouldBe`, it solves part of #80: the test suite won't crash anymore when displaying the diff if an lazy exception is stored in the input string. That's a good thing !
- Cool thing, user can write their own `shouldBe` with a different test algorithm and different timeout, so I think that's a good enough solution in order to close #92
- I introduced a `DEPRECATED` pattern synonym for `NotEqualButShouldHaveBeenEqual` so in theory, this is not a breaking change. Do you think we should keep it (keeps "cruft") or remove it (breaking change).

I still don't know what to do with:

```
-- | An assertion that says two 'String's should have been equal according to `==`.
--
-- Note that using function could mess up the colours in your terminal if the Texts contain ANSI codes.
-- In that case you may want to `show` your values first or use `shouldBe` instead.
stringsNotEqualButShouldHaveBeenEqual :: String -> String -> Assertion
stringsNotEqualButShouldHaveBeenEqual actual expected = NotEqualButShouldHaveBeenEqual actual expected

-- | An assertion that says two 'Text's should have been equal according to `==`.
--
-- Note that using function could mess up the colours in your terminal if the Texts contain ANSI codes.
-- In that case you may want to `show` your values first or use `shouldBe` instead.
textsNotEqualButShouldHaveBeenEqual :: Text -> Text -> Assertion
textsNotEqualButShouldHaveBeenEqual actual expected = NotEqualButShouldHaveBeenEqual (T.unpack actual) (T.unpack expected)

-- | An assertion that says two 'ByteString's should have been equal according to `==`.
bytestringsNotEqualButShouldHaveBeenEqual :: ByteString -> ByteString -> Assertion
bytestringsNotEqualButShouldHaveBeenEqual actual expected = NotEqualButShouldHaveBeenEqual (show actual) (show expected)
```

They are "pure" function, so I cannot compute the diff in it (so for now, they are still taking hours). I'll fix their usage location manually in the codebase, but what should we do for the exported API? Change them to `IO Assertion`, or introduce a deprecation warning?

Give me some feedbacks and I'll improve / edit

**Comment by NorfairKing** [comment]
Hey! I don't have much time right now but I saw this and got really excited so I wanted to respond quickly:

> I introduced a DEPRECATED pattern synonym for NotEqualButShouldHaveBeenEqual so in theory, this is not a breaking change. Do you think we should keep it (keeps "cruft") or remove it (breaking change).

I'm fine with breaking backward compatibility for this change.

> They are "pure" function, so I cannot compute the diff in it (so for now, they are still taking hours). I'll fix their usage location manually in the codebase, but what should we do for the exported API? Change them to IO Assertion, or introduce a deprecation warning?

I think we can turn them all into `_ -> _ -> IO Assertion` so that they each compute the diff as they go.
I also don't really care if the timeout is configurable.
It's not like you don't see the assertion, it's just the big diffs that you don't see then


I'm really happy we're getting around to doing this!
If it turns out I'm not being clear then I'll get back to writing a response when I've had some coffee :D

**Comment by guibou** [comment]
> I'm fine with breaking backward compatibility for this change.

I've removed the `DEPRECATED` function and patterns, it is way cleaner like that. I've updated the `CHANGELOG.md` to discuss the changes.

> I think we can turn them all into _ -> _ -> IO Assertion so that they each compute the diff as they go.

Done.

> I also don't really care if the timeout is configurable.

Done. Actually, the code for the timeout and diff computation is now "overridable" by the user (they can implement their own `shouldBe` and `goldenTestCompare`, so they can change the timeout or the diff algorithm as they please.

> It's not like you don't see the assertion, it's just the big diffs that you don't see then

Indeed.

I've cleaned the code, squashed, added a nice commit message, edited `Changelog.md` and `nix flake check` is running. I think it is now up to you ;)

Note: even if I'm doing that with my personal github account, let's assume that this work in done on behalf of novainsilico.ai and hence help the company comply with your software licence.

File: sydtest-test/test/Test/Syd/TimingSpec.hs [hunk 2]
```diff
@@ -28,6 +32,31 @@ spec = doNotRandomiseExecutionOrder $ do
...
+          case res of
+            Right () -> fail "The subtest must fail"
+            Left (_ :: ExitCode) -> pure ()
+
+    -- We timeout after 10s. If the test timeouts by itself, it means that the
+    -- diff timeout worked after 2seconds.
+    --
+    -- Why 10s? Because it's more than 2s, and gives enough room for scheduling
+    -- and additional operations (such as generating the random numbers)
+    -- without too much risk of generating a flaky test.
```
[the remark this answers is not in the document]
Comment (guibou) [reply sydtest-test/test/Test/Syd/TimingSpec.hs:58]: Thank you ;)

File: sydtest/src/Test/Syd/Output.hs [hunk 3]
```diff
@@ -463,8 +469,27 @@ splitChunksIntoLines =
 
 outputEqualityAssertionFailed :: String -> String -> [[Chunk]]
 outputEqualityAssertionFailed actual expected =
-  let diff = V.toList $ getTextDiff (T.pack actual) (T.pack expected)
-      -- Add a header to a list of lines of chunks
+  let diffM = unsafePerformIO $ do
```
[the remark this answers is not in the document]
Comment (guibou) [reply sydtest/src/Test/Syd/Output.hs]: Done as you suggested, smart design. Thank you the idea.

**Comment by guibou** [comment]
(nix flake check fails because the nixpkgs does not contains the latest version of `fast-myers-diff`, I'm bumping nixpkgs and reruning it.)

**Comment by guibou** [comment]
@NorfairKing the `nix flake checks` just fail because of some unit test which seems unrelated to the current changes. I've tried to add a commit https://github.com/NorfairKing/sydtest/pull/94/commits/64d34d8bb0fd61f786fc895e7d7c604292054294 in order to add some missing golden files.

**Comment by NorfairKing** [comment]
Rebased on top of the other changes.
Now waiting for https://staging.nix-ci.com/account/repo/5f9f1cce-5b39-4651-879c-58dd8ed3d3f7/suite/7e054c22-f412-4b2b-8e4b-a1da88c69055
to pass and then I'll release this as `sydtest-0.18.0.0`.

**Comment by NorfairKing** [comment]
@guibou The test failure was a true positive, `:sydtest-output-test` just needed a `--golden-reset`.

**Comment by NorfairKing** [comment]
merged manually and released.

**Pull request closed without merging**